File: python/2020/august/aug-10/main.py

# * Daily Coding Problem August 10 2020

# * [Easy] -- Uber

# * On election day, a voting machine writes data in the form
# * (voter_id, candidate_id) to a text file. Write a program that
# * reads this file as a stream and returns the top 3 candidates at
# * any given time. If you find a voter voting more than once, report this as fraud.

import pytest


# * Processes the votes from the file path up to n votes,
# * if n is not provided all votes are processed
def processVotes(filePath, n=None):
    file = open(filePath, "r")
    counts = {}
    voters = {}
    totalVotes = 0

    for line in file:
        # * Parse out the voter_id and candidate_id from line of text
        voter, candidate = line[1:len(line) - 2].split(", ")

        # * Check for fraud
        if voter in voters:
            print(
                f"Election Fraud! It appears that voter #{voter} has already cast their ballot")
            raise SystemExit(1)

        # * Register the vote and print out current state
        voters[voter] = True
        counts[candidate] = counts.setdefault(candidate, 0) + 1
        totalVotes += 1
        printVotes(totalVotes, counts)

        # * If n was provided and we have processed n votes, stop and return the top three
        if n and totalVotes == n:
            return topThreeCandidates(counts)

    # * Close out file and return the top three
    file.close()
    return topThreeCandidates(counts)


# * Returns the top three current candidates
def topThreeCandidates(candidates):
    result = []
    maxCandidates = 3
    currentCandidate = 0

    sorted_by_votes = {k: v for k, v in sorted(
        candidates.items(), key=lambda item: item[1], reverse=True)}

    for candidate, votes in sorted_by_votes.items():
        if currentCandidate >= maxCandidates:
            break
        currentCandidate += 1
        result.append(candidate)
    return result


# * Prints current voting state
def printVotes(totalVotes, counts):
    print(f"Votes: {totalVotes}\nTop 3 in order are: \n")
    topThree = topThreeCandidates(counts)
    for i, candidate in enumerate(topThree):
        print(
            f"{i + 1}: Candidate #{candidate} -- ({counts[candidate]}) votes")
    print("\n--------------------\n")


# * _______________________________TESTS___________________________


def test_process_votes_after_5_votes():
    assert processVotes("votes.txt", 5) == ['1', '2', '3']


def test_process_votes_after_3_votes():
    assert processVotes("votes.txt", 3) == ['1']


def test_process_votes_after_all_votes():
    assert processVotes("votes.txt") == ['1', '2', '3']


def test_process_votes_fraud(capsys):
    with pytest.raises(SystemExit) as exc:
        processVotes("votes_fraud.txt")
    assert exc.value.code == 1


# * ___________________ MAIN _______________________________

def main():
    print("Election day and the votes are ...\n-----------------------------------\n")
    processVotes("votes.txt")


main()

File: rust/2020/august/aug-10/src/main.rs
// * Daily Coding Problem August 10 2020

// * [Easy] -- Uber

// * On election day, a voting machine writes data in the form
// * (voter_id, candidate_id) to a text file. Write a program that
// * reads this file as a stream and returns the top 3 candidates at
// * any given time. If you find a voter voting more than once, report this as fraud.

use std::collections::{HashMap, HashSet};
use std::fs::File;
use std::io::{self, BufRead};
use std::path::Path;

// The output is wrapped in a Result to allow matching on errors
// Returns an Iterator to the Reader of the lines of the file.
fn read_lines<P>(filename: P) -> io::Result<io::Lines<io::BufReader<File>>>
where
    P: AsRef<Path>,
{
    let file = File::open(filename)?;
    Ok(io::BufReader::new(file).lines())
}

fn process_votes(votes_file: &str) {
    let mut rankings: HashMap<String, u32> = HashMap::new();
    let mut voters: HashSet<String> = HashSet::new();

    if let Ok(lines) = read_lines(votes_file) {
        for line in lines {
            if let Ok(mut content) = line {
                // * Stripping off the parens
                content.remove(0);
                content.remove(content.len() - 1);
                let vote: Vec<&str> = content.split(", ").collect();
                // * Check for fraud
                if voters.contains(vote[0]) {
                    panic!("Voter fraud detected!!\n It appears that voter {} has already cast their ballot.", vote[0]);
                } else {
                    voters.insert(String::from(vote[0]));
                }

                // * Cast their vote
                match &rankings.get(vote[1]) {
                    Some(vote_count) => rankings.insert(String::from(vote[1]), *vote_count + 1),
                    None => rankings.insert(String::from(vote[1]), 1),
                };
                println!("{}", content);
            }
        }
        let winner = rankings.iter_mut().max_by(|a, b| a.1.cmp(&b.1)).unwrap();
        println!("{:?}", winner);
        // rankings.remove_entry(winner.0);
        // let runner_up = rankings.iter_mut().max_by(|a, b| a.1.cmp(&b.1)).unwrap();
        // println!("{:?}", runner_up);
        // rankings.remove_entry(runner_up.0);

        // let third_place = rankings.iter_mut().max_by(|a, b| a.1.cmp(&b.1)).unwrap();
        // println!("{:?}", third_place);
        // rankings.remove_entry(third_place.0);
    }
}

fn main() {
    println!("It's Election Day and the votes are in!\n------------------\n");
    process_votes("votes.txt");
}
